crates/sui-rpc-api/proto/sui/rpc/v2alpha/effects.proto

```diff
@@ -0,0 +1,153 @@
+
+  // Information about the gas object. Also present in the `changed_objects` vector.
+  //
+  // System transaction that don't require gas will leave this as `None`.
+  optional ChangedObject gas_object = 8;
+
+  // The digest of the events emitted during execution,
+  // can be `None` if the transaction does not emit any event.
+  optional string events_digest = 9;
+
+  // The set of transaction digests this transaction depends on.
+  repeated string dependencies = 10;
+
+  // The version number of all the written Move objects by this transaction.
+  optional uint64 lamport_version = 11;
+
+  // Objects whose state are changed by this transaction.
+  repeated ChangedObject changed_objects = 12;
+
+  // Shared objects that are not mutated in this transaction. Unlike owned objects,
+  // read-only shared objects' version are not committed in the transaction,
+  // and in order for a node to catch up and execute it without consensus sequencing,
+  // the version needs to be committed in the effects.
+  repeated UnchangedSharedObject unchanged_shared_objects = 13;
+
+  // Auxiliary data that are not protocol-critical, generated as part of the effects but are stored separately.
+  // Storing it separately allows us to avoid bloating the effects with data that are not critical.
+  // It also provides more flexibility on the format and type of the data.
+  optional string auxiliary_data_digest = 14;
+}
+
+// Input/output state of an object that was changed during execution.
+message ChangedObject {
+  // ID of the object.
+  optional string object_id = 1;
+
+  enum InputObjectState {
+    INPUT_OBJECT_STATE_UNKNOWN = 0;
+    INPUT_OBJECT_STATE_DOES_NOT_EXIST = 1;
+    INPUT_OBJECT_STATE_EXISTS = 2;
+  }
+
+  optional InputObjectState input_state = 2;
+
+  // Version of the object before this transaction executed.
+  optional uint64 input_version = 3;
+  // Digest of the object before this transaction executed.
+  optional string input_digest = 4;
+  // Owner of the object before this transaction executed.
+  optional Owner input_owner = 5;
+
+  enum OutputObjectState {
+    OUTPUT_OBJECT_STATE_UNKNOWN = 0;
+    OUTPUT_OBJECT_STATE_DOES_NOT_EXIST = 1;
+    OUTPUT_OBJECT_STATE_OBJECT_WRITE = 2;
+    OUTPUT_OBJECT_STATE_PACKAGE_WRITE = 3;
+  }
+
+  optional OutputObjectState output_state = 6;
+  // Version of the object after this transaction executed.
+  optional uint64 output_version = 7;
+  // Digest of the object after this transaction executed.
+  optional string output_digest = 8;
+  // Owner of the object after this transaction executed.
+  optional Owner output_owner = 9;
+
+  enum IdOperation {
+    ID_OPERATION_UNKNOWN = 0;
+    ID_OPERATION_NONE = 1;
+    ID_OPERATION_CREATED = 2;
+    ID_OPERATION_DELETED = 3;
+  }
+
+  // What happened to an `ObjectId` during execution.
+  optional IdOperation id_operation = 10;
+
+  // Type information is not provided by the effects structure but is instead
+  // provided by an indexing layer
```

Comment:
What's the pruning story for this data? Will this work by trying to fetch the object at this version and getting its type and/or trying to get the live version of the object and fetching the type there (which is fine to do today)? Either case, in the absence of an implicit KV store fallback, how often will we be able to correctly return the type of an object?

It would be good to document some of the nuance about that here given this field is the odd one out among the rest of the effects structure.

Reply:
I think the semantics of how the field is populated is going to be dependent on the implementation and not as relevant for how the schema is populated. I called this out just to indicate that today it's not a part of the core effects structure (maybe it will be later though). Either way, similar to balance changes at least in the node, my plan was to be to have a table for this calculated at execution time so that we can trivially load this without loading the objects and it would be pruned when the txn is pruned.